Importing Dependencies

In [ ]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

In [ ]:
from sklearn.preprocessing import StandardScaler

In [ ]:
# loading the dataset to pandas dataframe

credit_card_data=pd.read_csv('/content/creditcard.csv')

In [ ]:
#first five rows of the dataset

credit_card_data.head()

Time        V1        V2        V3        V4        V5        V6        V7  \
0   0.0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1   0.0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2   1.0 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3   1.0 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4   2.0 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9  ...       V21       V22       V23       V24       V25  \
0  0.098698  0.363787  ... -0.018307  0.277838 -0.110474  0.066928  0.128539   
1  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288 -0.339846  0.167170   
2  0.247676 -1.514654  ...  0.247998  0.771679  0.909412 -0.689281 -0.327642   
3  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321 -1.175575  0.647376   
4 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458  0.141267 -0.206010   

        V26       V27       V28  Amount  Class  
0 -0.189115  0.133558 -0.021053  149.62      0  
1  0.125895 -0.008983  0.014724    2.69      0  
2 -0.139097 -0.055353 -0.059752  378.66      0  
3 -0.221929  0.062723  0.061458  123.50      0  
4  0.502292  0.219422  0.215153   69.99      0  

[5 rows x 31 columns]

In [ ]:
credit_card_data.tail()

Time         V1         V2        V3        V4        V5  \
284802  172786.0 -11.881118  10.071785 -9.834783 -2.066656 -5.364473   
284803  172787.0  -0.732789  -0.055080  2.035030 -0.738589  0.868229   
284804  172788.0   1.919565  -0.301254 -3.249640 -0.557828  2.630515   
284805  172788.0  -0.240440   0.530483  0.702510  0.689799 -0.377961   
284806  172792.0  -0.533413  -0.189733  0.703337 -0.506271 -0.012546   

              V6        V7        V8        V9  ...       V21       V22  \
284802 -2.606837 -4.918215  7.305334  1.914428  ...  0.213454  0.111864   
284803  1.058415  0.024330  0.294869  0.584800  ...  0.214205  0.924384   
284804  3.031260 -0.296827  0.708417  0.432454  ...  0.232045  0.578229   
284805  0.623708 -0.686180  0.679145  0.392087  ...  0.265245  0.800049   
284806 -0.649617  1.577006 -0.414650  0.486180  ...  0.261057  0.643078   

             V23       V24       V25       V26       V27       V28  Amount  \
284802  1.014480 -0.509348  1.436807  0.250034  0.943651  0.823731    0.77   
284803  0.012463 -1.016226 -0.606624 -0.395255  0.068472 -0.053527   24.79   
284804 -0.037501  0.640134  0.265745 -0.087371  0.004455 -0.026561   67.88   
284805 -0.163298  0.123205 -0.569159  0.546668  0.108821  0.104533   10.00   
284806  0.376777  0.008797 -0.473649 -0.818267 -0.002415  0.013649  217.00   

        Class  
284802      0  
284803      0  
284804      0  
284805      0  
284806      0  

[5 rows x 31 columns]

In [ ]:
credit_card_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 284807 entries, 0 to 284806
Data columns (total 31 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   Time    284807 non-null  float64
 1   V1      284807 non-null  float64
 2   V2      284807 non-null  float64
 3   V3      284807 non-null  float64
 4   V4      284807 non-null  float64
 5   V5      284807 non-null  float64
 6   V6      284807 non-null  float64
 7   V7      284807 non-null  float64
 8   V8      284807 non-null  float64
 9   V9      284807 non-null  float64
 10  V10     284807 non-null  float64
 11  V11     284807 non-null  float64
 12  V12     284807 non-null  float64
 13  V13     284807 non-null  float64
 14  V14     284807 non-null  float64
 15  V15     284807 non-null  float64
 16  V16     284807 non-null  float64
 17  V17     284807 non-null  float64
 18  V18     284807 non-null  float64
 19  V19     284807 non-null  float64
 20  V20     284807 non-null  float64
 21  V21     28

In [ ]:
#checking number of missing values in each column

credit_card_data.isnull().sum()

Time      0
V1        0
V2        0
V3        0
V4        0
V5        0
V6        0
V7        0
V8        0
V9        0
V10       0
V11       0
V12       0
V13       0
V14       0
V15       0
V16       0
V17       0
V18       0
V19       0
V20       0
V21       0
V22       0
V23       0
V24       0
V25       0
V26       0
V27       0
V28       0
Amount    0
Class     0
dtype: int64

In [ ]:
#distribution of legit transaction and fraudulent transanctions

credit_card_data['Class'].value_counts()

Class
0    284315
1       492
Name: count, dtype: int64

This dataset is highly unbalanced

0 -> Normal Transactions
1 -> Fraud Transactions

In [ ]:
#separating the data for analysis

legit= credit_card_data[credit_card_data.Class == 0]
fraud= credit_card_data[credit_card_data.Class == 1]

In [ ]:
print(legit.shape)
print(fraud.shape)

(284315, 31)
(492, 31)


In [ ]:
#statistical measures of the data

legit.Amount.describe()

count    284315.000000
mean         88.291022
std         250.105092
min           0.000000
25%           5.650000
50%          22.000000
75%          77.050000
max       25691.160000
Name: Amount, dtype: float64

In [ ]:
fraud.Amount.describe()

count     492.000000
mean      122.211321
std       256.683288
min         0.000000
25%         1.000000
50%         9.250000
75%       105.890000
max      2125.870000
Name: Amount, dtype: float64

In [ ]:
#compare the values for both the transactions
credit_card_data.groupby('Class').mean()

Time        V1        V2        V3        V4        V5  \
Class                                                                   
0      94838.202258  0.008258 -0.006271  0.012171 -0.007860  0.005453   
1      80746.806911 -4.771948  3.623778 -7.033281  4.542029 -3.151225   

             V6        V7        V8        V9  ...       V20       V21  \
Class                                          ...                       
0      0.002419  0.009637 -0.000987  0.004467  ... -0.000644 -0.001235   
1     -1.397737 -5.568731  0.570636 -2.581123  ...  0.372319  0.713588   

            V22       V23       V24       V25       V26       V27       V28  \
Class                                                                         
0     -0.000024  0.000070  0.000182 -0.000072 -0.000089 -0.000295 -0.000131   
1      0.014049 -0.040308 -0.105130  0.041449  0.051648  0.170575  0.075667   

           Amount  
Class              
0       88.291022  
1      122.211321  

[2 rows x 30 columns]

<h3>Under-Sampling</h3> as  284315 for legit transanctions whereas only 492 for fraudulent transanctions


Build a sample dataset containing similar distribution of normal transanctions and fraudulent transanctions

Number of Fraudulent Transanctions ---> 492

In [ ]:
legit_sample= legit.sample(n=492)

Concatenating two dataframes

In [ ]:
new_dataset=pd.concat([legit_sample,fraud],axis=0)

In [ ]:
new_dataset.head()

Time        V1        V2        V3        V4        V5        V6  \
227212  144980.0  2.183071 -1.956649  0.068715 -1.170917 -2.295550 -0.199056   
246667  153274.0 -2.242649  2.466863 -1.597996 -3.539564  4.912658  2.910610   
276552  167164.0 -1.878219 -2.531252  2.167856 -2.101045  0.567219 -1.581044   
262750  160631.0 -0.298025 -0.371161 -0.327879 -2.679069 -1.228783  0.814183   
169825  119876.0 -1.061347  0.767103  1.917264  1.133565  0.202898  0.110280   

              V7        V8        V9  ...       V21       V22       V23  \
227212 -1.961341  0.189191  0.182335  ... -0.378232 -0.499138  0.360845   
246667  3.693151 -2.269575  4.549397  ... -0.940561  0.810360 -0.559691   
276552 -1.593195  0.185554 -1.871317  ... -0.166894 -0.783605  0.317221   
262750  1.211660 -0.361247 -2.671722  ...  0.124486  0.885209 -0.271698   
169825  0.535542  0.128700  0.166003  ... -0.562436 -1.306638 -0.164613   

             V24       V25       V26       V27       V28  Amount  Class  
227212 -0.199937 -0.545472 -0.286052  0.055407 -0.036687   32.00      0  
246667  0.688312  0.177690 -0.217602 -0.206728 -1.888349    0.77      0  
276552 -0.094249  0.287253 -0.317074 -0.017846  0.107115   59.00      0  
262750  0.132869 -0.017981  0.089837  0.052139  0.054062  297.00      0  
169825 -0.152151  0.371415 -0.784955  0.265158  0.135952   45.19      0  

[5 rows x 31 columns]

In [ ]:
new_dataset['Class'].value_counts()

Class
0    492
1    492
Name: count, dtype: int64

In [ ]:
new_dataset.groupby('Class').mean()

Time        V1        V2        V3        V4        V5  \
Class                                                                   
0      92726.934959 -0.021385  0.104356  0.018886 -0.112878 -0.083066   
1      80746.806911 -4.771948  3.623778 -7.033281  4.542029 -3.151225   

             V6        V7        V8        V9  ...       V20       V21  \
Class                                          ...                       
0     -0.003550  0.047827  0.047817  0.048955  ...  0.021293 -0.038580   
1     -1.397737 -5.568731  0.570636 -2.581123  ...  0.372319  0.713588   

            V22       V23       V24       V25       V26       V27       V28  \
Class                                                                         
0     -0.050404  0.030274 -0.067973 -0.027410 -0.001739 -0.004798  0.008882   
1      0.014049 -0.040308 -0.105130  0.041449  0.051648  0.170575  0.075667   

           Amount  
Class              
0       89.337256  
1      122.211321  

[2 rows x 30 columns]

Splitting the data into Features and Targets

In [ ]:
X=new_dataset.drop(columns='Class',axis=1)
Y=new_dataset['Class'].astype(int)

In [ ]:
print(X)

            Time        V1        V2        V3        V4        V5        V6  \
227212  144980.0  2.183071 -1.956649  0.068715 -1.170917 -2.295550 -0.199056   
246667  153274.0 -2.242649  2.466863 -1.597996 -3.539564  4.912658  2.910610   
276552  167164.0 -1.878219 -2.531252  2.167856 -2.101045  0.567219 -1.581044   
262750  160631.0 -0.298025 -0.371161 -0.327879 -2.679069 -1.228783  0.814183   
169825  119876.0 -1.061347  0.767103  1.917264  1.133565  0.202898  0.110280   
...          ...       ...       ...       ...       ...       ...       ...   
279863  169142.0 -1.927883  1.125653 -4.518331  1.749293 -1.566487 -2.010494   
280143  169347.0  1.378559  1.289381 -5.004247  1.411850  0.442581 -1.326536   
280149  169351.0 -0.676143  1.126366 -2.213700  0.468308 -1.120541 -0.003346   
281144  169966.0 -3.113832  0.585864 -5.399730  1.817092 -0.840618 -2.943548   
281674  170348.0  1.991976  0.158476 -2.583441  0.408670  1.151147 -0.096695   

              V7        V8        V9  .

In [ ]:
print(Y)

227212    0
246667    0
276552    0
262750    0
169825    0
         ..
279863    1
280143    1
280149    1
281144    1
281674    1
Name: Class, Length: 984, dtype: int64


Split the data into Training and Testing Data

In [ ]:
X_train,X_test,Y_train,Y_test=train_test_split(X,Y,test_size=0.2,stratify=Y,random_state=2)

In [ ]:
print(X.shape,X_train.shape,X_test.shape)

(984, 30) (787, 30) (197, 30)


Model Training

Logistic Regression

In [ ]:
model=LogisticRegression(max_iter=20000)

In [ ]:
#training the logistic regression model with training data

model.fit(X_train, Y_train)

LogisticRegression(max_iter=20000)

Model Evaluation

Accuracy Score

In [ ]:
#accuracy on training data

X_train_prediction=model.predict(X_train)
training_data_accuracy=accuracy_score(X_train_prediction,Y_train)

In [ ]:
print('Accuracy on training data: ',training_data_accuracy*100)

Accuracy on training data:  96.06099110546378


In [ ]:
#accuracy on testing data

X_test_prediction= model.predict(X_test)
test_data_accuracy=accuracy_score(X_test_prediction, Y_test)

In [ ]:
print('Accuracy on training data: ',test_data_accuracy*100)

Accuracy on training data:  92.89340101522842
